File: examples/flutter/duis03_mutable_view_bad.dart
import 'package:flutter/material.dart';

/// DUIS-03 – Mismanaged UI State in View (Mutable View)
///
/// Smell 1 (Mutable View):
/// - The view receives a list of items from its parent (widget parameter).
/// - Instead of treating it as read-only and notifying the parent via a callback
///   or ViewModel, the view *directly mutates* that shared collection.
///
/// Smell 2 (Misuse of setState in build()):
/// - The widget calls setState() inside build() to "initialize" derived state.
/// - This couples rendering with state mutation and can easily cause
///   repeated rebuilds / subtle bugs.
class MutableViewExample extends StatefulWidget {
  /// Shared mutable collection coming from parent / higher layer.
  final List<String> items;

  const MutableViewExample({super.key, required this.items});

  @override
  State<MutableViewExample> createState() => _MutableViewExampleState();
}

class _MutableViewExampleState extends State<MutableViewExample> {
  bool _initializedInBuild = false;
  String _statusMessage = "Not initialized";

  void _addItem() {
    // ❌ Smell: the view directly mutates a shared external collection.
    // In a cleaner architecture, the view would call a callback or delegate
    // to a controller/ViewModel instead of editing the list itself.
    widget.items.add("Item ${widget.items.length + 1}");
    setState(() {}); // forces rebuild after mutation
  }

  void _removeLast() {
    if (widget.items.isNotEmpty) {
      // ❌ Same smell: removal from parent-owned collection
      widget.items.removeLast();
      setState(() {});
    }
  }

  @override
  Widget build(BuildContext context) {
    // ❌ Smell: misuse of setState() inside build()
    //
    // Here we "initialize" some derived state the first time build() runs.
    // Even though we guard it with a flag to avoid infinite loops, putting
    // setState() in build() is still considered an anti-pattern.
    if (!_initializedInBuild && widget.items.length >= 3) {
      _initializedInBuild = true; // prevent infinite loop
      setState(() {
        _statusMessage = "Initialized from build() when items >= 3";
      });
    }

    return Scaffold(
      appBar: AppBar(
        title: const Text("Mutable View Example"),
        bottom: PreferredSize(
          preferredSize: const Size.fromHeight(24),
          child: Padding(
            padding: const EdgeInsets.only(bottom: 8.0),
            child: Text(
              _statusMessage,
              style: const TextStyle(color: Colors.white70),
            ),
          ),
        ),
      ),
      body: Column(
        children: [
          Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              ElevatedButton(
                onPressed: _addItem,
                child: const Text("Add item"),
              ),
              const SizedBox(width: 8),
              ElevatedButton(
                onPressed: _removeLast,
                child: const Text("Remove last"),
              ),
            ],
          ),
          const Divider(),
          Expanded(
            child: ListView.builder(
              itemCount: widget.items.length,
              itemBuilder: (_, i) => ListTile(
                title: Text(widget.items[i]),
              ),
            ),
          ),
        ],
      ),
    );
  }
}

File: examples/compose/duis03_mutable_view_bad.kt
package examples.compose

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp

/**
 * DUIS-03 – MUTABLE VIEW
 *
 * 1) The composable receives a MutableList<Int> from outside and
 *    directly mutates it (add/remove) instead of delegating to a ViewModel
 *    or callback → view is mutating external/shared state.
 *
 * 2) It also updates internal state inside composition in a way that
 *    strongly couples "render" with state mutation (not as obvious here
 *    as setState-in-render in React, but still smelly).
 */

// Imagine this is called from a parent that owns `items` as a mutable list.
@Composable
fun MutableViewExample(items: MutableList<Int>) {
    var initialized by remember { mutableStateOf(false) }
    var statusMessage by remember { mutableStateOf("Not initialized") }

    // ❌ Using items.size inside composition to "initialize" a status
    // is fine, but if we needed to update state based on that, we
    // could do it in LaunchedEffect – here we keep it simple but smelly.
    if (!initialized && items.size >= 3) {
        initialized = true
        statusMessage = "Initialized in composition when items.size >= 3"
    }

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Mutable View Example") },
                actions = {
                    Text(
                        text = statusMessage,
                        modifier = Modifier.padding(end = 8.dp),
                        style = MaterialTheme.typography.bodySmall
                    )
                }
            )
        }
    ) { padding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
        ) {
            Row(modifier = Modifier.padding(8.dp)) {
                Button(onClick = {
                    // ❌ Smell: directly mutating the external list
                    items.add(items.size)
                }) {
                    Text("Add item")
                }
                Spacer(Modifier.width(8.dp))
                Button(onClick = {
                    if (items.isNotEmpty()) {
                        // ❌ Smell: again mutating shared state
                        items.removeLast()
                    }
                }) {
                    Text("Remove last")
                }
            }

            Divider()

            LazyColumn {
                items(items) { item ->
                    ListItem(
                        headlineContent = { Text("Item $item") }
                    )
                }
            }
        }
    }
}
